extensions/binding/org.eclipse.smarthome.binding.mqtt.generic/src/main/java/org/eclipse/smarthome/binding/mqtt/generic/internal/discovery/Homie300Discovery.java

```diff
@@ -0,0 +1,152 @@
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link Homie300Discovery} is responsible for discovering device nodes that follow the
+ * Homie 3.x convention (https://github.com/homieiot/convention).
+ *
+ * @author David Graeff - Initial contribution
+ */
+@Component(immediate = true, service = DiscoveryService.class, configurationPid = "discovery.mqtthomie")
+@NonNullByDefault
+public class Homie300Discovery extends AbstractDiscoveryService implements TopicDiscovered {
+    private final Logger logger = LoggerFactory.getLogger(Homie300Discovery.class);
+    protected @Nullable MqttTopicDiscovery mqttTopicDiscovery;
+
+    public Homie300Discovery() {
+        super(Stream.of(MqttBindingConstants.HOMIE300_MQTT_THING).collect(Collectors.toSet()), 500, true);
+    }
+
+    @Override
+    @Activate
+    protected void activate(@Nullable Map<String, @Nullable Object> config) {
+        super.activate(config);
+    };
+
+    @Override
+    @Deactivate
+    protected void deactivate() {
+        unsetThingRegistry(null);
+        super.deactivate();
+    }
+
+    @Reference
+    public void setThingRegistry(ThingRegistry service) {
+        mqttTopicDiscovery = new MqttTopicDiscovery(service, this, "homie/+/$homie");
+    }
+
+    public void unsetThingRegistry(@Nullable ThingRegistry service) {
+        if (mqttTopicDiscovery != null) {
+            mqttTopicDiscovery.stopBackgroundDiscovery();
+            mqttTopicDiscovery = null;
+        }
+    }
+
+    @Override
+    protected void startScan() {
+        if (mqttTopicDiscovery != null) {
+            mqttTopicDiscovery.startScan();
+        }
+        stopScan();
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        if (mqttTopicDiscovery != null) {
+            mqttTopicDiscovery.startBackgroundDiscovery();
+        }
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        if (mqttTopicDiscovery != null) {
+            mqttTopicDiscovery.stopBackgroundDiscovery();
+        }
+    }
+
+    /**
+     * @param topic A topic like "homie/mydevice/$homie"
+     * @return Returns the "mydevice" part of the example
+     */
+    public static @Nullable String extractDeviceID(String topic) {
+        String[] strings = topic.split("/");
+        if (strings.length > 2) {
+            return strings[1];
+        }
+        return null;
+    }
+
+    /**
+     * Returns true if the version is something like "3.x". We accept
+     * version 3 up to but not including version 4 of the homie spec.
+     */
+    public static boolean checkVersion(String versionString) {
+        String[] strings = versionString.split("\\.");
+        if (strings.length < 2) {
+            return false;
+        }
+        return strings[0].equals("3");
```

Comment:
`"3".equals(strings[0])` or simply replace all with `return versionString.startsWith("3.")`?

Reply:
The way of how the specification of this convention works is rather unusual. They say they follow the SemVer schema, but regarding structure parsing breakage, they only consider "clients" (end devices), not the "hosts". In the future I might need to check for an allowed range of minor versions as well. That's why I didn't want to use `startsWith` here. 

const string first for the comparison can be done of course.